### PR TITLE
Validate geometries on update/create

### PR DESCRIPTION
**Pull request by tsauerwein**
Validate the geometries on update and create using the PostGIS functions `ST_IsSimple` and `ST_IsValid`.

### CODE REVIEW

File: c2cgeoportal/views/layers.py [hunk 1]
```diff
@@ -244,8 +245,16 @@ def security_cb(r, feature, o):
...
 
-        protocol = self._get_protocol_for_layer(layer, before_create=security_cb)
-        return protocol.create(self.request)
+                # check if geometry is valid
+                self.validate_geometry_(spatial_elt)
+
+        protocol = self._get_protocol_for_layer(layer, before_create=check_geometry)
+        try:
+            features = protocol.create(self.request)
+            return features
```
Comment (asaunier) [inline c2cgeoportal/views/layers.py]: Is this intermediary variable "features" really necessary?

Comment (tsauerwein) [inline c2cgeoportal/views/layers.py]: It's not necessary, but I wanted to make it clear that `protocol.create()` returns a list of features.

File: c2cgeoportal/views/layers.py [hunk 2]
```diff
@@ -283,8 +293,27 @@ def security_cb(r, feature, o):
...
 
-        protocol = self._get_protocol_for_layer(layer, before_update=security_cb)
-        return protocol.update(self.request, feature_id)
+            # check is geometry is valid
+            self.validate_geometry_(spatial_elt)
+
+        protocol = self._get_protocol_for_layer(layer, before_update=check_geometry)
+        try:
+            feature = protocol.update(self.request, feature_id)
+            return feature
```
Comment (asaunier) [inline c2cgeoportal/views/layers.py]: Same question about the intermediary variable.

Comment (asaunier) [inline c2cgeoportal/views/layers.py]: Or maybe for the readability of the code? (showing explicitly what is returned here)

**Comment by asaunier** [comment]
Looks good to me!

File: c2cgeoportal/views/layers.py (hunk 2 above)
Comment (tsauerwein) [inline c2cgeoportal/views/layers.py]: Same reason. :) I had to open the source code of `papyrus` to know what `protocol.update()` returns. This variable should make it a bit clearer.

Comment (asaunier) [inline c2cgeoportal/views/layers.py]: OK, that makes sense!

File: c2cgeoportal/views/layers.py [hunk 3]
```diff
@@ -283,8 +293,27 @@ def security_cb(r, feature, o):
...
+        except TopologicalError, e:
+            self.request.response.status_int = 400
+            return {'validation_error': str(e)}
+
+    def validate_geometry_(self, geom):
```
Comment (sbrunner) [inline c2cgeoportal/views/layers.py]: Why this name ?
To mark it as private it should be prefixed by an underscore, not postfixed ...
https://docs.python.org/2/tutorial/classes.html#private-variables-and-class-local-references

File: vars.yaml [hunk 4]
```diff
@@ -14,6 +14,7 @@ vars:
         "role_id" "^-?[0-9]*$"'
     mapfile_data_subselect: "SELECT ST_Collect(ra.area) FROM main.restrictionarea AS ra, main.role_restrictionarea AS rra, main.layer_restrictionarea AS lra, main.treeitem AS la WHERE rra.role_id = %role_id% AND rra.restrictionarea_id = ra.id AND lra.restrictionarea_id = ra.id AND lra.layer_id = la.id AND la.name ="
     fqdn: facter fqdn
+    geometry_validation: False
```
Comment (sbrunner) [inline vars.yaml]: I think that should be true by default :-)

File: c2cgeoportal/views/layers.py (hunk 3 above)
Comment (tsauerwein) [inline c2cgeoportal/views/layers.py]: Sorry, I confused this with the ol3 code style.

**Comment by sbrunner** [comment]
Missing `prwgma: nocover` 
https://coveralls.io/files/447823921#L312

**Comment by tsauerwein** [comment]
> Missing prwgma: nocover 

So that the coverage does not decrease? You are cheating! ;)

**Comment by sbrunner** [comment]
Yes I'm cheating :-)

More seriously, for me it's acceptable to don't test all exceptions cases, and is we don't use the `pragma: nocover` It's difficult to see it code that should be tested is relay tested :-)

**Comment by tsauerwein** [comment]
Ok, I see your point.

https://github.com/camptocamp/c2cgeoportal/pull/1203

**Pull request merged**